File: packages/gbnf-experiment/data/20260909T151143Z_170b737d/ported_implementation/tests/unit/print.test.ts
import assert from "node:assert/strict";
import { describe, it } from "node:test";

import { RuleChar } from "../../src/grammar_graph/grammar_graph_types.ts";
import type {
  PrintableNode,
  PrintablePointer,
  PrintOpts,
  UnresolvedRule,
} from "../../src/grammar_graph/grammar_graph_types.ts";
import { GraphNode } from "../../src/grammar_graph/graph_node.ts";
import { GraphPointer } from "../../src/grammar_graph/graph_pointer.ts";
import { print_graph_node, print_graph_pointer } from "../../src/grammar_graph/print.ts";
import { RuleRef } from "../../src/grammar_graph/rule_ref.ts";

const COLORS: Record<string, string> = {
  "\x1b[34m": "BLUE",
  "\x1b[36m": "CYAN",
  "\x1b[32m": "GREEN",
  "\x1b[31m": "RED",
  "\x1b[90m": "GRAY",
  "\x1b[33m": "YELLOW",
};

const mock_colorize = (text: string | number, color: string): string => {
  if (!(color in COLORS)) {
    throw new Error(`Invalid color: ${color}`);
  }
  return `[${COLORS[color]}]:${text}`;
};

const create_mock_node = (
  id: string,
  rule: UnresolvedRule,
  next: PrintableNode | null = null,
): PrintableNode => ({
  id,
  rule,
  next,
  print: () => `Node(${id})`,
});

const create_mock_graph_pointer = (node: PrintableNode): PrintablePointer => ({
  node,
  print: () => `Pointer to ${node.id}`,
});

describe("print_graph_pointer", () => {
  it("prints graph pointer details", () => {
    // A pointer whose parent chain is 1,1,1 renders that id between the marker.
    const parent = new GraphPointer(
      new GraphNode(new RuleChar([97]), { stackId: 1, pathId: 1, stepId: 1 }),
    );
    const pointer = new GraphPointer(
      new GraphNode(new RuleChar([98]), { stackId: 2, pathId: 2, stepId: 2 }),
      parent,
    );

    const result = print_graph_pointer(pointer)({
      colorize: mock_colorize,
      pointers: [],
      show_position: false,
    });

    assert.equal(result, "[RED]:*[RED]:1,1,1");
  });

  it("prints just the marker when there is no parent", () => {
    const pointer = new GraphPointer(
      new GraphNode(new RuleChar([97]), { stackId: 1, pathId: 1, stepId: 1 }),
    );
    assert.equal(
      print_graph_pointer(pointer)({ colorize: mock_colorize }),
      "[RED]:*",
    );
  });
});

describe("print_graph_node", () => {
  it("prints a graph node with a character rule", () => {
    const mock_node = create_mock_node("1", new RuleChar([65]));
    const result = print_graph_node(mock_node)({
      colorize: mock_colorize,
      show_position: false,
      pointers: [],
    });
    assert.equal(result, "[GRAY]:[[YELLOW]:A[GRAY]:]");
  });

  it("prints a graph node with a rule reference", () => {
    const mock_node = create_mock_node("1", new RuleRef(200));
    const result = print_graph_node(mock_node)({
      colorize: mock_colorize,
      show_position: true,
      pointers: [],
    });
    assert.equal(result, "[BLUE]:{[GRAY]:1[BLUE]:}[GRAY]:Ref([GREEN]:200[GRAY]:)");
  });

  it("renders ranges and newlines", () => {
    const mock_node = create_mock_node("1", new RuleChar([[97, 122], 10]));
    const result = print_graph_node(mock_node)({
      colorize: mock_colorize,
      show_position: false,
    });
    assert.equal(result, "[GRAY]:[[YELLOW]:[YELLOW]:a[YELLOW]:z\\n[GRAY]:]");
  });

  it("appends the next node after an arrow", () => {
    const next = create_mock_node("2", new RuleChar([66]));
    const mock_node = create_mock_node("1", new RuleChar([65]), next);
    const opts: PrintOpts = { colorize: mock_colorize, show_position: false };
    assert.equal(
      print_graph_node(mock_node)(opts),
      "[GRAY]:[[YELLOW]:A[GRAY]:][GRAY]:-> Node(2)",
    );
  });

  it("annotates pointers that sit on the node", () => {
    const mock_node = create_mock_node("1", new RuleChar([65]));
    const result = print_graph_node(mock_node)({
      colorize: mock_colorize,
      show_position: false,
      pointers: [create_mock_graph_pointer(mock_node)],
    });
    assert.equal(
      result,
      "[GRAY]:[[YELLOW]:A[GRAY]:][GRAY]:[[YELLOW]:Pointer to 1[GRAY]:]",
    );
  });
});

File: proposed-projects/round-trip-experiment/reverse/20260909T002206Z_273cf6b9/20260910T042804Z_e72f6c04/ported_implementation/gbnf/grammar_graph/print_test.py
import pytest

from . import print as module
from .grammar_graph_types import PrintOpts, RuleChar
from .print import print_graph_node, print_graph_pointer
from .rule_ref import RuleRef

COLORS = {
    "\x1b[34m": "BLUE",
    "\x1b[36m": "CYAN",
    "\x1b[32m": "GREEN",
    "\x1b[31m": "RED",
    "\x1b[90m": "GRAY",
    "\x1b[33m": "YELLOW",
}


def mock_colorize(text, color):
    if color not in COLORS:
        raise ValueError(f"Invalid color: {color}")
    return f"[{COLORS[color]}]:{text}"


class MockNode:
    def __init__(self, id, rule, next=None):
        self.id = str(id)
        self.rule = rule
        self.next = next

    def print(self, opts):
        return f"Node({self.id})"


class MockGraphPointer:
    def __init__(self, node):
        self.node = node
        self.parent = None

    def print(self, opts):
        return f"Pointer to {self.node.id}"


@pytest.fixture(autouse=True)
def stub_get_parent_stack_id(monkeypatch):
    monkeypatch.setattr(module, "get_parent_stack_id", lambda pointer, col: "")
    yield monkeypatch


def test_print_graph_pointer_prints_graph_pointer_details_correctly(
    stub_get_parent_stack_id,
):
    mock_node = MockNode(1, RuleRef(100))
    mock_pointer = MockGraphPointer(mock_node)
    stub_get_parent_stack_id.setattr(
        module, "get_parent_stack_id", lambda pointer, col: "foo"
    )
    result = print_graph_pointer(mock_pointer)(
        PrintOpts(colorize=mock_colorize, pointers=None, show_position=False)
    )
    assert result == "[RED]:*foo"


def test_print_graph_node_prints_graph_node_with_a_character_rule():
    mock_node = MockNode(1, RuleChar([65]))
    result = print_graph_node(mock_node)(
        PrintOpts(colorize=mock_colorize, show_position=False, pointers=None)
    )
    assert result == "[GRAY]:[[YELLOW]:A[GRAY]:]"


def test_print_graph_node_prints_graph_node_with_a_rule_reference():
    mock_node = MockNode(1, RuleRef(200))
    result = print_graph_node(mock_node)(
        PrintOpts(colorize=mock_colorize, show_position=True, pointers=None)
    )
    assert result == "[BLUE]:{[GRAY]:1[BLUE]:}[GRAY]:Ref([GREEN]:200[GRAY]:)"
